### ISSUE TITLE
Porting R channel to conda-forge

### ISSUE THREAD

**Issue by johanneskoester** [conda-forge/staged-recipes#2009]
This issue will track the effort to port the R channel into conda-forge. Please feel free to join if you like.

# How to contribute
We can now start migrating R packages from conda/conda-recipes. 

* The recipes below are sorted topologically, into levels. 
* Each level has to be completed before we can do the next, because the next level will depend on recipes from the previous.
* Go to the currently tackled level, choose a subset of the recipes, and create a pull request with them. * Please mention the PR in this issue and tick the checkbox once the PR has been merged.
* When creating recipes, please use conda-build >=2.1.9 for an up to date skeleton.
* Add `skip: true  # [win32]`.
* Remove irrelevant comments.
* Add yourself, `johanneskoester` and `bgruening` as a maintainer.
* See my cleanup script below.

# Prerequisites
* [x] java-jdk/openjdk (PR #2002)
* [x] r-base (PR #2047)
* [x] r-boot (PR #2098)
* [x] dependencies of r-recommended
* [x] r-recommended and r metapackages (PR #2780)

# R packages to migrate (topological sort levels such that all r-package depencencies are satisfied)
# Level 1
* [x] r-acepack (PR #2767)
* [x] r-adgoftest (PR #2767)
* [x] r-assertthat (PR #2767)
* [x] r-backports (PR #2767)
* [x] r-base64enc (PR #2767)
* [x] r-bdsmatrix (PR #2767)
* [x] r-bh (PR #2767)
* [x] r-bit (PR #2767)
* [x] r-bitops (PR #2767)
* [x] r-boot
* [x] r-brew (PR #2781)
* [x] r-cairo (PR #2766)
* [x] r-chron (PR #2781)
* [x] r-cluster
* [x] r-codetools
* [x] r-colorspace (PR #2781)
* [x] r-crayon
* [x] r-ctv (PR #2781)
* [x] r-curl (PR #2781)
* [x] r-data.table (PR #2781)
* [x] r-dbi (PR #2781)
* [x] r-deoptimr (PR #2781)
* [x] r-dichromat (PR #2781)
* [x] r-digest
* [x] r-estimability (PR #2782)
* [x] r-findpython (PR #2766)
* [x] r-foreign
* [x] r-formatr (PR #2782)
* [x] r-formula (PR #2782)
* [x] r-fracdiff (PR #2782)
* [x] r-functional (PR #2782)
* [x] r-getopt (PR #2766)
* [x] r-git2r (PR #2782)
* [x] r-gridbase (PR #2782)
* [x] r-gsl (PR #2783)
* [x] r-gsw
* [x] r-gtable (PR #2783)
* [x] r-gtools (PR #2783)
* [x] r-highr (PR #2783)
* [x] r-hms (PR #2783)
* [x] r-inline (PR #2766)
* [x] r-iterators (PR #2783)
* [x] r-janeaustenr (PR #2783)
* [x] r-jpeg (PR #2783)
* [x] r-jsonlite
* [x] r-kernsmooth
* [x] r-labeling (PR #2784)
* [x] r-lars (PR #2784)
* [x] r-lattice
* [x] r-lazyeval (PR #2784)
* [x] r-leaps (PR #2784)
* [x] r-logging (PR #2784)
* [x] r-magrittr
* [x] r-manipulate (PR #2784)
* [x] r-maps (PR #2784)
* [x] r-mass
* [x] r-mime (PR #2784)
* [x] r-mnormt (PR #2784)
* [x] r-modeltools (PR #2784)
* [x] r-mvtnorm (PR #2784)
* [x] r-nloptr (PR #2784)
* [x] r-nlp (PR #2784)
* [x] r-nnet
* [x] r-numderiv (PR #2784)
* [x] r-openssl (PR #2860)
* [x] r-packrat (PR #2812)
* [x] r-perm (PR #2812)
* [x] r-plogr (PR #2812)
* [x] r-png (PR #2812)
* [x] r-praise (PR #2766)
* [x] r-profilemodel (PR #2799)
* [x] r-proto (PR #2766)
* [x] r-pspline (PR #2799)
* [x] r-quadprog (PR #2799)
* [x] r-r.methodss3 (PR #2799)
* [x] r-r6
* [x] r-randomforest (PR #2799)
* [x] r-rappdirs (PR #2799)
* [x] r-rcolorbrewer (PR #2799)
* [x] r-rcpp (PR #2787)
* [x] r-registry (PR #2799)
* [x] r-repr
* [x] r-rjava  (PR #2860)
* [x] r-rjson (PR #2766)
* [x] r-rjsonio (PR #2770)
* [x] r-rodbc (PR #2770)
* [x] r-rpart
* [x] r-rserve (PR #2770)
* [x] r-rstudioapi (PR #2770)
* [x] r-rzmq (PR #2770)
* [x] r-slam (PR #2770)
* [x] r-snowballc (PR #2770)
* [x] r-sourcetools (PR #2770)
* [x] r-sparsem (PR #2770)
* [x] r-spatial
* [x] r-stabledist (PR #2770)
* [x] r-stanheaders (PR #2769)
* [x] r-stringdist (PR #2769)
* [x] r-stringi
* [x] r-testit (PR #2769)
* [x] r-timedate (PR #2769)
* [x] r-uuid
* [x] r-vgam (PR #2753)
* [x] r-viridislite (PR #2769)
* [x] r-whisker (PR #2769)
* [x] r-withr (PR #2769)
* [x] r-xml (PR #2769)
* [x] r-xtable (PR #2769)
* [x] r-yaml (PR #2769)
* [x] r-glmnet (PR #2890)
* [x] r-grpreg (PR #2890)
* [x] r-foreach (PR #2890)
# Level 2
* [x] r-argparse (PR #2890)
* [x] r-bestglm (PR #2890)
* [x] r-bit64 (PR #2891)
* [x] r-brglm (PR #2891)
* [x] r-catools (PR #2891)
* [x] r-class
* [x] r-coda (PR #2891)
* [x] r-config (PR #2891)
* [x] r-gdata (PR #2891)
* [x] r-gridextra (PR #2892)
* [x] r-hexbin (PR #2892)
* [x] r-htmltools (PR #2892)
* [x] r-httpuv (PR #2892)
* [x] r-httr (PR #2892)
* [x] r-hunspell (PR #2892)
* [x] r-irdisplay
* [x] r-latticeextra (PR #2893)
* [x] r-mapproj (PR #2893)
* [x] r-markdown (PR #2893)
* [x] r-matrix
* [x] r-memoise (PR #2893)
* [x] r-minqa (PR #2893)
* [x] r-modelmetrics (PR #2893)
* [x] r-munsell (PR #2893)
* [x] r-nlme
* [x] r-oce
* [x] r-pbdzmq
* [x] r-pcapp (PR #2895)
* [x] r-pki (PR #2895)
* [x] r-plyr (PR #2895)
* [x] r-psych (PR #2895)
* [x] r-r.oo (PR #2895)
* [x] r-rcpparmadillo (PR #2895)
* [x] r-rcurl (PR #2895)
* [x] r-readxl (PR #2896)
* [x] r-rex (PR #2896)
* [x] r-rhive (PR #2896)
* [x] r-rlist (PR #2896)
* [x] r-robustbase (PR #2896)
* [x] r-rprojroot (PR #2896)
* [x] r-sp (PR #2896)
* [x] r-stringr
* [x] r-tensorflow (PR #2898)
* [x] r-testthat (PR #2898)
* [x] r-tibble (PR #2898)
* [x] r-tm (PR #2898)
* [x] r-tokenizers (PR #2898)
* [x] r-xlsxjars (PR #2898)
* [x] r-xml2 (PR #2898)
* [x] r-zoo (PR #2753)
* [x] r-commonmark (PR #2912)
* [x] r-desc (PR #2912)
# Level 3
* [x] r-ape (PR #2906)
* [x] r-copula (PR #2906)
* [x] r-cvtools (PR #2906)
* [x] r-devtools (PR #2906)
* [x] r-domc (PR #2906)
* [x] r-doparallel (PR #2906)
* [x] r-dplyr (PR #2906)
* [x] r-e1071 (PR #2906)
* [x] r-evaluate
* [x] r-feather (PR #2907)
* [x] r-forcats (PR #2907)
* [x] r-gplots (PR #2907)
* [x] r-htmlwidgets (PR #2907)
* [x] r-irlba (PR #2907)
* [x] r-kohonen (PR #2907)
* [x] r-lmtest (PR #2907)
* [x] r-lubridate (PR #2907)
* [x] r-maptools (PR #2907)
* [x] r-matrixmodels (PR #2907)
* [x] r-mgcv
* [x] r-nycflights13 (PR #2908)
* [x] r-pkgmaker (PR #2908)
* [x] r-proc (PR #2908)
* [x] r-pryr (PR #2772)
* [x] r-r.utils (PR #2908)
* [x] r-raster (PR #2908)
* [x] r-rcppeigen (PR #2908)
* [x] r-readr (PR #2908)
* [x] r-reshape (PR #2908)
* [x] r-reshape2
* [x] r-roxygen2 (PR #2912)
* [x] r-rsconnect (PR #2909)
* [x] r-rsqlite (PR #2909)
* [x] r-rversions (PR #2909)
* [x] r-sandwich (PR #2909)
* [x] r-scales (PR #2909)
* [x] r-seacarb (PR #2909)
* [x] r-selectr (PR #2909)
* [x] r-shiny (PR #2912)
* [x] r-survival
* [x] r-tilegramsr (PR #2910)
* [x] r-tseries (PR #2910)
* [x] r-urca (PR #2910)
* [x] r-weatherdata (PR #2910)
* [x] r-xlsx (PR #2910)
* [x] r-xts (PR #2910)
* [x] r-sf (PR #2910)
* [x] r-units (PR #2910)
# Level 4
* [x] r-dt (PR #2915)
* [x] r-dygraphs (PR #2915)
* [x] r-fwdselect (PR #2915)
* [x] r-ggplot2 (PR #2915)
* [x] r-ggvis (PR #2915)
* [x] r-haven (PR #2915)
* [x] r-irkernel
* [x] r-knitr (PR #2915)
* [x] r-lahman (PR #2915)
* [x] r-leaflet (PR #2915)
* [x] r-lme4 (PR #2915)
* [x] r-miniui (PR #2915)
* [x] r-viridis (PR #2915)
* [x] r-purrr (PR #2916)
* [x] r-quantreg (PR #2916)
* [x] r-rmr2 (PR #2916)
* [x] r-rngtools (PR #2916)
* [x] r-rocr (PR #2916)
* [x] r-rvest (PR #2916)
* [x] r-shinybs (PR #2916)
* [x] r-shinydashboard (PR #2916)
* [x] r-shinysky (PR #2916)
* [x] r-shinythemes (PR #2918)
* [x] r-sparklyr (PR #2918)
* [x] r-strucchange (PR #2918)
* [x] r-th.data (PR #2918)
* [x] r-threejs (PR #2918)
* [x] r-tidyr (PR #2918)
* [x] r-ttr (PR #2918)
* [x] r-visnetwork (PR #2918)
* [x] r-car (PR #2921)
* [x] r-caret (PR #2921)
* [x] r-gmp (PR #2921)
* [x] r-matrixstats (PR #2921)
* [x] r-qdaptools (PR #2921)
* [x] r-tidyverse (PR #2921)
# Level 5
* [x] r-anomalydetection (PR #2921)
* [x] r-bradleyterry2 (PR #2921)
* [x] r-broom (PR #2921)
* [x] r-forecast (PR #2921)
* [x] r-htmltable (PR #2921)
* [x] r-microbenchmark (PR #2921)
* [x] r-mlmrev (PR #2921)
* [x] r-multcomp (PR #2921)
* [x] r-nmf  (PR #2920)
* [x] r-pbkrtest  (PR #2920)
* [x] r-quantmod  (PR #2920)
* [x] r-rgl  (PR #2920)
* [x] r-rmarkdown  (PR #2920)
* [x] r-rstan  (PR #2920)
* [x] r-shinyjs  (PR #2920)
* [x] r-vars  (PR #2920)
# Level 6
* [x] r-coin (PR #2923)
* [x] r-formattable (PR #2923)
* [x] r-gistr (PR #2923)
* [x] r-hmisc (PR #2923)
* [x] r-igraph (PR #2923)
* [x] r-lsmeans (PR #2923)
* [x] r-modelr (PR #2923)
* [x] r-tidytext (PR #2923)
# Level 7
* [x] r-aer (PR #2924)
* [x] r-afex (PR #2924)
* [x] r-highcharter (PR #2924)
* [x] r-influencer (PR #2924)
* [x] r-lintr (PR #2924)
* [x] r-networkd3 (PR #2924)
* [x] r-plm (PR #2924)
* [x] r-rbokeh (PR #2924)
# Level 8
* [x] r-diagrammer (PR #2929)
* [x] r-essentials (PR #2929)

# Cleanup script:
In order to clean up recipes created with `conda skeleton cran`, I apply the following script:

```
#!/bin/sh

RECIPE=$1/meta.yaml
BUILDSH=$1/build.sh
BUILDBAT=$1/bld.bat

sed -i -e '/^ *#.*$/d' $RECIPE
sed -i -e '/^$/N;/^\n$/D' $RECIPE

echo extra: >> $RECIPE
echo "  recipe-maintainers:" >> $RECIPE
echo "    - johanneskoester" >> $RECIPE
echo "    - bgruening" >> $RECIPE

echo '"%R%" CMD INSTALL --build .' > $BUILDBAT
echo 'if errorlevel 1 exit 1' >> $BUILDBAT

echo '#!/bin/bash' > $BUILDSH
echo '' >> $BUILDSH
echo '$R CMD INSTALL --build .' >> $BUILDSH
```

**Comment by asmeurer** [comment]
Wow, this is a great effort. I hope you can do it. I did all the original R packaging. I don't have the resources nowadays to help out with this, but if you have a question you think I might know the answer to feel free to ping me.

**Comment by johanneskoester** [comment]
Thanks a lot @asmeurer!

**Comment by johanneskoester** [comment]
@conda-forge/core could you please direct people that provide PRs for R packages to this issue? I'd like to keep track of what is already done.

**Comment by jakirkham** [comment]
Does the list include things like `r-oce` ( https://github.com/conda-forge/staged-recipes/pull/2461 ), `r-mass` ( https://github.com/conda-forge/staged-recipes/pull/2666 ), `r-cluster` ( https://github.com/conda-forge/staged-recipes/pull/2666 ), `r-class` ( https://github.com/conda-forge/staged-recipes/pull/2666 ), `r-codetools`
 ( https://github.com/conda-forge/staged-recipes/pull/2668 ), `r-foreign`
 ( https://github.com/conda-forge/staged-recipes/pull/2668 ), `r-kernsmooth`
 ( https://github.com/conda-forge/staged-recipes/pull/2668 ), `r-matrix`
 ( https://github.com/conda-forge/staged-recipes/pull/2668 ), `r-mgcv` ( https://github.com/conda-forge/staged-recipes/pull/2672 ), `r-nlme` ( https://github.com/conda-forge/staged-recipes/pull/2672 ), `r-nnet` ( https://github.com/conda-forge/staged-recipes/pull/2672 ), `r-rpart` ( https://github.com/conda-forge/staged-recipes/pull/2672 ), `r-spatial` ( https://github.com/conda-forge/staged-recipes/pull/2672 ), `r-crayon`  ( https://github.com/conda-forge/staged-recipes/pull/2460 ), `r-digest`  ( https://github.com/conda-forge/staged-recipes/pull/2460 ), `r-jsonlite`  ( https://github.com/conda-forge/staged-recipes/pull/2460 ), `r-lattice`  ( https://github.com/conda-forge/staged-recipes/pull/2460 ), `r-magrittr`  ( https://github.com/conda-forge/staged-recipes/pull/2460 ), `r-r6`  ( https://github.com/conda-forge/staged-recipes/pull/2460 ), `r-repr`  ( https://github.com/conda-forge/staged-recipes/pull/2460 ), `r-stringi`  ( https://github.com/conda-forge/staged-recipes/pull/2460 ), or `r-uuid`  ( https://github.com/conda-forge/staged-recipes/pull/2460 )?

**Comment by johanneskoester** [comment]
The list only includes packages that are not done yet.

**Comment by jakirkham** [comment]
Cool, wanted to make sure that (a) they were listed here somehow and (b) check with you that they had been noted somehow.

**Comment by jakirkham** [comment]
Also PR ( https://github.com/conda-forge/staged-recipes/pull/1597 ) has a bunch of incomplete recipes that are up for grabs if anyone wants to see them through.

**Comment by jrderuiter** [comment]
Who should we reference as recipe maintainers? I'm happy to help port a number of packages (https://github.com/conda-forge/staged-recipes/pull/2772), but do not necessarily want to be responsible for maintaining them.

**Comment by johanneskoester** [comment]
@johanneskoester and @bgruening

**Comment by bgruening** [comment]
Thanks @jrderuiter!

**Comment by jrderuiter** [comment]
Just wondering, what is the rationale for skipping win32? (Not that I miss it, just curious.)

**Comment by johanneskoester** [comment]
We don't have an  openjdk build for win32. It is not offered by Zulu (the provider) and we don't build it ourselves so far.

**Comment by jrderuiter** [comment]
r-nycflights13 should be moved to level 3, as it import r-tibble.

**Comment by johanneskoester** [comment]
You are right. This dependency has apparently changes since 0.2.0, on which my dependency tree was based. Let's hope this is the only case.

**Comment by jakirkham** [comment]
Updated the level 1 listing after seeing some feedstocks get added this morning. Appears most sailed through. Also looks like that part of the list is nearly completed. 🎉

However, `r-rzmq` appeared to have some issues with CI registration. Am trying to readd the existing recipes to staged-recipes to see if that fixes the issue. ( https://github.com/conda-forge/staged-recipes/pull/2889 )

**Comment by jakirkham** [comment]
Yep, re-feedstocking `r-rzmq` through staged-recipes seems to have done the trick. All CIs ran and passed. Packages for all supported platforms uploaded. Checked the box in the list above.

**Comment by bgruening** [comment]
Level 2 should be finished if the above PR gets merged, currently blocking a few failing feedstock creations. xref: https://github.com/conda-forge/staged-recipes/pull/2897

**Comment by johanneskoester** [comment]
You mean level 1?

**Comment by bgruening** [comment]
No level 2:  #2890 #2891 #2892 #2893 #2895 #2896 #2898

**Comment by jrderuiter** [comment]
Great! :) I'll try to help with levels 3 and 4 if I have some free time this week.

**Comment by johanneskoester** [comment]
Oh cool!! Haven't seen that. I guess it is not possible for somebody else than me to add the PR numbers to the list at the top, right? I'll update it. Maybe mention me once you add a new PR, so that we don't duplicate work.

**Comment by bgruening** [comment]
@johanneskoester will do this. But I think we need to wait for now to get the failing feedstock from Level 1 in. Enjoy your weekend!

**Comment by bgruening** [comment]
I have created some hacks to create these recipes.
Just put the package name in `packages.txt` in the form of `r-foobar` and run the `./fix.sh` script. You will end up with a bunch of `r-*` folders. Please check them manually. Especially the LICENSE section. This sections should not contain the word `license`.

Hope this helps a little bit.

[build_helpers.tar.gz](https://github.com/conda-forge/staged-recipes/files/966083/build_helpers.tar.gz)

**Comment by isuruf** [comment]
Here are some of the packages that have failed.
- [ ] r-assertthat
- [ ] r-bestglm
- [ ] r-cellranger
- [ ] r-ctv
- [ ] r-functional
- [ ] r-glmnet
- [ ] r-labeling
- [ ] r-rappdirs
- [ ] r-readxl
- [ ] r-tensorflow

Please restart the CI for these repos. Most of them are network failures or package missing failures due to race conditions.

@jakirkham, I found this by reverting https://github.com/conda-forge/conda-forge.github.io/commit/7ede26729a882f32641a543dfa18022d3cd219bc and generating locally. Can we restore CI badges in a separate webpage so that we can see what packages are failing? This would be especially useful for @johanneskoester and @bgruening to maintain the huge number of r packages.

**Comment by bgruening** [comment]
* [ ] [r-assertthat](https://github.com/conda-forge/r-assertthat-feedstock)
* [ ] [r-bestglm](https://github.com/conda-forge/r-bestglm-feedstock)
* [ ] [r-cellranger](https://github.com/conda-forge/r-cellranger-feedstock)
* [ ] [r-ctv](https://github.com/conda-forge/r-ctv-feedstock)
* [ ] [r-functional](https://github.com/conda-forge/r-functional-feedstock)
* [ ] [r-glmnet](https://github.com/conda-forge/r-glmnet-feedstock)
* [ ] [r-labeling](https://github.com/conda-forge/r-labeling-feedstock)
* [ ] [r-rappdirs](https://github.com/conda-forge/r-rappdirs-feedstock)
* [ ] [r-readxl](https://github.com/conda-forge/r-readx-feedstock)
* [ ] [r-tensorflow](https://github.com/conda-forge/r-tensorflow-feedstock)

**Comment by bgruening** [comment]
Besides https://travis-ci.org/conda-forge/r-readxl-feedstock and https://travis-ci.org/conda-forge/r-cellranger-feedstock everything should be fine now. Can someone please restart these, I can't for unknown reason.

**Comment by bgruening** [comment]
Work on Level3 started in https://github.com/conda-forge/staged-recipes/pull/2906 and https://github.com/conda-forge/staged-recipes/pull/2907

**Comment by lpantano** [comment]
I am starting with 

r-rsconnect
r-rsqlite
r-rversions
r-sandwich
r-scales
r-seacarb
r-selectr

**Comment by bgruening** [comment]
@lanzagar cool thanks! Here is a new batch: https://github.com/conda-forge/staged-recipes/pull/2908

**Comment by bgruening** [comment]
I will do the last ones:


r-tilegramsr
 r-tseries
 r-urca
 r-weatherdata
 r-xlsx
 r-xts

**Comment by bgruening** [comment]
I have pushed the latest version of my cleanup-scripts to https://github.com/bgruening/conda_r_skeleton_helper.

**Comment by lpantano** [comment]
r-commonmark 
r-desc 

at https://github.com/conda-forge/staged-recipes/issues/2912

**Comment by bgruening** [comment]
Level 4 is here: #2915 #2916 #2918

**Comment by johanneskoester** [comment]
@bgruening you are crazy :-). Thanks a lot!

**Comment by bgruening** [comment]
Level 5 is here: https://github.com/conda-forge/staged-recipes/pull/2920 & https://github.com/conda-forge/staged-recipes/pull/2921

**Comment by jrderuiter** [comment]
Awesome! Are there similar plans for migrating bioconductor packages? Or will they remain in bioconda for now?

**Comment by johanneskoester** [comment]
We only port non - bio stuff to conda-forge. Hence bioconductor will stay part of bioconda.

**Comment by jakirkham** [comment]
Checked off some more stuff. Looks like a couple things got turned around in the topological ordering. In particular `r-tidyverse` -> `r-modelr` -> `r-broom`. However `r-broom` is built now. Have restarted the Travis CI build for `r-modelr`. Though someone will need to do the same for the Travis CI build of `r-tidyverse` as well after the build of `r-modelr` completes.

**Comment by johanneskoester** [comment]
Thanks! The problem is that the toposort was done on the recipes from the R channel, which were sometimes older versions with slighly different dependencies. When creating the new recipes for our PRs we naturally received newer versions, sometimes breaking the toposort. Luckily, there are only few such cases.

**Comment by jakirkham** [comment]
Thanks for the details @johanneskoester. AFAIAC there will inevitably be a few issues when porting this number of interdependent packages. I'm not too worried about it personally. Just trying to help catch issues as they arise.

**Comment by jakirkham** [comment]
> @jakirkham, I found this by reverting conda-forge/conda-forge.github.io@7ede267 and generating locally. Can we restore CI badges in a separate webpage so that we can see what packages are failing?

@isuruf, we decided to remove the CI badges some time ago as it made loading the webpage very slow and we were also concerned it had negative effects for our service providers. Not to mention they were sometimes out-of-date. So I don't think we will add that back. I expect the need for this sort of thing will diminish once the porting is finished. Though am open to other suggestions of how we manage such things for future porting efforts.

xref: https://github.com/conda-forge/conda-forge.github.io/pull/210#issuecomment-241140417

**Comment by bgruening** [comment]
@jakirkham I was able to get the new order and most of it is solved. I was not updating the list from above as I do not have write access to this issue. However, it does not matter we are finished this week.

**Comment by johanneskoester** [comment]
@isuruf since you seem to be active at the moment, I'll ask you directly: if we want to update the R version in the future (via CONDA_R), what is the mechanism we are supposed to apply? I guess there is some central place where CONDA_R is defined, and when we set a new version, conda-smithy will create corresponding pull requests to all r-related feedstocks?

**Comment by isuruf** [comment]
I think you just need to update `r-base` and the maintenance bots will send PRs for updating.

**Comment by johanneskoester** [comment]
So you mean they will automatically figure out that an r-base version update implies that CONDA_R shall be updated?

**Comment by isuruf** [comment]
Yes, it will create a version matrix and there'll be 2 builds for the 2 versions. If you want to drop support for the old version, conda-smithy will have to be updated. @jakirkham, correct me if I'm wrong.

**Comment by johanneskoester** [comment]
We are done! Thanks to everybody who helped!

**Comment by mjsteinbaugh** [comment]
You all are amazing. Already done before I had a chance to start some pull requests. Thanks for all the hard work.

**Comment by jakirkham** [comment]
@johanneskoester:

> > So you mean they will automatically figure out that an r-base version update implies that CONDA_R shall be updated?

@isuruf:

> Yes, it will create a version matrix and there'll be 2 builds for the 2 versions. If you want to drop support for the old version, conda-smithy will have to be updated.

Correct. This is the [relevant line]( https://github.com/conda-forge/conda-smithy/blob/d5684d3a8d01878abde84d887bdd5a3a9bab2814/conda_smithy/configure_feedstock.py#L629 ) in `conda-smithy`. Similar lines exist in staged-recipes as well.

**Comment by johanneskoester** [comment]
So what if we want to pin to CONDA_R=3.4 instead of CONDA_R=3.4.0?

**Comment by johanneskoester** [comment]
Thanks, @mjsteinbaugh

**Comment by jakirkham** [comment]
> So what if we want to pin to CONDA_R=3.4 instead of CONDA_R=3.4.0?

When are you thinking of doing this?

**Comment by johanneskoester** [comment]
Within the next weeks. Unfortunately R 3.4.0 has been recently released. Now, it does not make too much sense to update everything in bioconda against 3.3.2 in conda-forge. Instead, we think it is best to build against 3.4 directly. However, one can already see that pinning to patch versions does not scale in the long run (both in terms of storage and build time). We have evaluated though that it should be fine to pin to the minor version (aka 3.4, not 3.4.0). R will print warnings if a package has been built with a newer version. But in general this is considered to work. If individual packages need special treatment, those feedstocks can always use a more specific CONDA_R.

**Comment by jakirkham** [comment]
TBH I'm not sure to what extent `conda-smithy`/`conda-build-all` have control over this. If `conda-build` just uses the value of `CONDA_R` verbatim, then we can probably add some short hack with a simple test to `conda-smithy` for the interim. If you want to give this a try, then I'll find some time to take a look. In the long term we should discuss a way to do this with `conda-build` 3. If it turns out `conda-build` is doing something more sophisticated, then we might need to just move everything to `conda-build` 3 first.

Part of the reason I say this is `conda-build` 3 is going to change a lot of stuff when it comes to pinning and doing matrix builds. Please take a look at PR ( https://github.com/conda/conda-docs/pull/414 ) for details if you haven't seen it yet. This will impact things like how we handle pinnings, matrix builds, and may even result in us dropping use of `conda-build-all`.

**Comment by asmeurer** [comment]
Curious what is your strategy to keep these feedstocks up-to-date? Is there a rerender-like script that can update them automatically?

**Comment by ocefpaf** [comment]
> Curious what is your strategy to keep these feedstocks up-to-date? Is there a rerender-like script that can update them automatically?

There is not such a thing yet. I do know that @mingwandroid triggers updates with `conda-skeleton`, is that correct @mingwandroid?

I had a GSoC student with a nice proposal to create a `conda-smithy update`, but he did not make NumFOCUS cut. Luckily he is still interested in implementing that. More soon...

**Comment by asmeurer** [comment]
Assumedly this wouldn't be too difficult to do reusing the existing rerender tooling.

**Comment by mingwandroid** [comment]
An old msys2 contributor had a tool to look for version updates that would modify the PKGBUILDS.

I find the dependencies change too a bit and there's some awfulness regarding build numbers for some old R packages in defaults that no longer see new versions that you need to be very careful about too (this was to do with get license fields in place). In some cases those packages may not be getting used anyone and there's an argument for dropping them.

Currently the process is automatic to recreate them all with conda skeleton but then each one is manually compared (using beyond compare).

I'd love to see something in place to figure out the most popular packages on CRAN and make recipes for them too.

**Comment by asmeurer** [comment]
Something that I had thought about doing back when I was maintaining the R packages, but never got around to, was somehow making it so that the base recipe was completely auto-generated from conda skeleton, and any modifications on top of that recipe (most commonly, additional build steps for Windows) would be in separate files. With this workflow, you could regenerate the recipes with conda skeleton, and keep the modifications on top of that separate (and, obviously, update skeleton itself as much as possible to keep what can be automated automated). This sort of thing might require some modifications to conda-build. I don't know if it's possible to merge two files to make a meta.yaml.

**Comment by mingwandroid** [comment]
Since I added the msys2 infrastructure to the Ananconda Distribution there's no longer anything custom about Windows, but the idea of isolating any custom stuff that there is to a something like a normal patch file is interesting.

**Comment by asmeurer** [comment]
> Since I added the msys2 infrastructure to the Ananconda Distribution there's no longer anything custom about Windows, but the idea of isolating any custom stuff that there is to something like a normal patch file is interesting.

Maybe R itself doesn't, but certain packages will. I remember some packages, especially those with C extensions, require "manual" steps to build on Windows. I am curious, though, which packages, if any, did require additional modification beyond skeleton. 

> .. then again, as a fan of git, I'd propose having those custom modifications in a specially named branch, then you'd just rebase that on top after each update and fix any conflicts. Maybe conda skeleton could help with that too though?

I don't like the idea of using git here. All the data to reproduce and build a recipe should be on the main git branch. 

> Merging two files to make a meta.yaml doesn't address build.sh / bld.bat.

Merging build.sh/bld.bat is easy, as those are scripting languages that can call out to other files. The skeleton generated build.sh could just have a check for "custom_build.sh" in the recipe and run that, and if it doesn't exist run the default build script.

**Comment by mingwandroid** [comment]
> Maybe R itself doesn't, but certain packages will. I remember some packages, especially those with C extensions, require "manual" steps to build on Windows. I am curious, though, which packages, if any, did require additional modification beyond skeleton.

Adding packages that require compilation requires a bit of work for all platforms but it's no worse and no more hacky on Windows than macOS or Linux.

**Comment by johanneskoester** [comment]
@jakirkham yes, as far as I know conda-build takes the value of CONDA_R verbatim. So if it is 3.4, it will just pin to 3.4 and all the patch versions should work. I would have thought that there is some way to tell conda-smithy: "Update all feedstocks with CONDA_R to this value". Is that not the case? If not, how do you update Python?

**Comment by jakirkham** [comment]
Good to know. Adding a new version isn't an issue that will happen once a new `r-base` is released and feedstocks are re-rendered. The issue is that `CONDA_R` has been pinned to the minor version and if we aren't going to do that anymore, we need to change the logic in `conda-smithy` and/or `conda-build-all`. Does that make sense?

**Comment by johanneskoester** [comment]
Yes, makes sense. So basically it should simply determine the latest r-base x.y.z and set CONDA_R to x.y.
